pkg/webhook/pod/mutating/pod_probe_marker.go

```diff
@@ -0,0 +1,101 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutating
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/util"
+	utilclient "github.com/openkruise/kruise/pkg/util/client"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+	"k8s.io/kube-openapi/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// mutating relate-pub annotation in pod
+func (h *PodCreateHandler) podProbeMakerMutatingPod(ctx context.Context, req admission.Request, pod *corev1.Pod) (skip bool, err error) {
+	if len(req.AdmissionRequest.SubResource) > 0 || req.AdmissionRequest.Operation != admissionv1.Create ||
+		req.AdmissionRequest.Resource.Resource != "pods" {
+		return true, nil
+	}
+	ppmList := &appsv1alpha1.PodProbeMarkerList{}
+	if err = h.Client.List(context.TODO(), ppmList, &client.ListOptions{Namespace: pod.Namespace}, utilclient.DisableDeepCopy); err != nil {
+		return false, err
+	} else if len(ppmList.Items) == 0 {
+		return true, nil
+	}
+
+	containers := sets.NewString()
+	for _, c := range pod.Spec.Containers {
+		containers.Insert(c.Name)
+	}
+	for _, c := range pod.Spec.InitContainers {
+		if util.IsRestartableInitContainer(&c) {
+			containers.Insert(c.Name)
+		}
+	}
+
+	matchedPodProbeMarkerName := sets.NewString()
+	matchedProbeKey := sets.NewString()
+	matchedConditions := sets.NewString()
+	matchedProbes := make([]appsv1alpha1.PodContainerProbe, 0)
+	for _, obj := range ppmList.Items {
+		// This error is irreversible, so continue
+		labelSelector, err := util.ValidatedLabelSelectorAsSelector(obj.Spec.Selector)
+		if err != nil {
+			continue
+		}
+		// If a PUB with a nil or empty selector creeps in, it should match nothing, not everything.
+		if labelSelector.Empty() || !labelSelector.Matches(labels.Set(pod.Labels)) {
+			continue
+		}
+		for i := range obj.Spec.Probes {
+			probe := obj.Spec.Probes[i]
+			key := fmt.Sprintf("%s/%s", probe.ContainerName, probe.Name)
+			if matchedConditions.Has(probe.PodConditionType) || matchedProbeKey.Has(key) || !containers.Has(probe.ContainerName) || probe.PodConditionType == "" {
+				continue
+			}
+			// No need to pass in marker related fields
+			probe.MarkerPolicy = nil
+			matchedProbes = append(matchedProbes, probe)
+			matchedProbeKey.Insert(key)
+			matchedConditions.Insert(probe.PodConditionType)
+			if !matchedPodProbeMarkerName.Has(obj.Name) {
```

Comment:
no need to check agaist the existence of entry, set operation will ensure only one entry exists